app/models/targeting.rb

```diff
@@ -0,0 +1,29 @@
+class Targeting < ActiveRecord::Base
+
+  belongs_to :user
+  belongs_to :bookmark
+
+  has_many :targeting_hosts, :dependent => :destroy
+  has_many :hosts, :through => :targeting_hosts
+  has_many :template_invocations, :dependent => :destroy
+
+  validate :bookmark_or_query
+
+  def resolve_hosts!
+    fail _("Cannot resolve hosts without a user") if user.nil?
+    fail _("Cannot resolve hosts without a bookmark or search query") if bookmark.nil? && search_query.blank?
+    query = bookmark.try(:query) || search_query
```

Comment:
I thought we had discussed that a user could provide a query or select a bookmark.  I guess we could store the bookmark's query as part of this (if they selected a bookmark).  is that what you meant?